relay-server/src/actors/upstream.rs

```diff
@@ -276,161 +292,226 @@ impl UpstreamRelay {
-                Ok(())
-            });
-        Box::new(future)
+            .into_actor(self)
+            .then(|send_result, slf, ctx| {
+                slf.handle_http_response_status(request, send_result, ctx)
+                    .into_actor(slf)
+            })
+            .spawn(ctx);
+    }
+
+    /// Checks what happened to the request and takes appropriate action
+    ///
+    /// - if the request was sent, notify the response sender
+    /// - if the request was not send:
+    ///     - if it was a network error
+    ///         - if request should be retried
+    ///             - schedule a retry
+    ///         - if there is no scheduled Authentication request
+    ///             - schedule authentication (with backoff)
+    ///     - if it was a non recoverable error, notify the response sender with the error
+    fn handle_http_response_status(
+        &mut self,
+        request: UpstreamRequest,
+        send_result: Result<ClientResponse, UpstreamRequestError>,
+        ctx: &mut Context<Self>,
+    ) -> impl Future<Item = (), Error = ()> {
+        if let Err(err) = send_result {
+            if err.is_network_error() {
+                self.handle_network_error(ctx);
+                if request.retry {
+                    self.enqueue(request, ctx, EnqueuePosition::Back);
+                }
+                return futures::future::failed(());
+            }
+            // we only retry network errors, forward this error and finish
+            request.response_sender.send(Err(err)).ok();
+        } else {
+            // reset any previously failed status
+            // TODO should we reset on success or on anything that is not a network error ?
+            self.first_error = None;
+            // success forward the result and finish
+            request.response_sender.send(send_result).ok();
+        }
+
+        futures::future::ok(())
+    }
+
+    fn enqueue(
+        &mut self,
+        request: UpstreamRequest,
+        ctx: &mut Context<Self>,
+        position: EnqueuePosition,
+    ) {
+        let push = match position {
+            EnqueuePosition::Front => VecDeque::push_front,
+            EnqueuePosition::Back => VecDeque::push_back,
+        };
+        let queue: Option<_>;
+
+        match request.priority {
+            RequestPriority::Immediate => {
+                // Immediate is special and bypasses the queue. Directly send the request and return
+                // the response channel rather than waiting for `PumpHttpMessageQueue`.
+                self.send_request(request, ctx);
+                return;
+            }
+            RequestPriority::Low => {
+                queue = Some(&mut self.low_prio_requests);
+            }
+            RequestPriority::High => {
+                queue = Some(&mut self.high_prio_requests);
+            }
+        };
+
+        if let Some(queue) = queue {
+            let name = request.priority.name();
+            push(queue, request);
```

Comment:
_nit_: Please inline the match on position for readability:

```suggestion
            match position {
                EnqueuePosition::Front => queue.push_front(request),
                EnqueuePosition::Back => queue.push_back(request),
            }
```

Reply:
ok